src/connection.js

```diff
@@ -0,0 +1,112 @@
+const { Map, fromJS } = require('immutable')
+const OpSet = require('./op_set')
+
+// Returns true if all components of clock1 are less than or equal to those of clock2.
+// Returns false if the clocks are incomparable or if clock1 is greater than clock2.
+function lessOrEqual(clock1, clock2) {
+  return clock1.keySeq().concat(clock2.keySeq()).reduce(
+    (result, key) => (result && clock1.get(key, 0) <= clock2.get(key, 0)),
+    true)
+}
+
+function mergeClock(clockMap, docId, clock) {
+  clock = clockMap.get(docId, Map()).mergeWith((x, y) => Math.max(x, y), clock)
+  return clockMap.set(docId, clock)
+}
+
+// Keeps track of the communication with one particular peer. Allows updates for many documents to
+// be multiplexed over a single connection.
+//
+// To integrate a connection with a particular networking stack, two functions are used:
+// * `sendMsg` (callback passed to the constructor, will be called when local state is updated)
+//   takes a message as argument, and sends it out to the remote peer.
+// * `receiveMsg` (method on the connection object) should be called by the network stack when a
+//   message is received from the remote peer.
+//
+// The documents to be synced are managed by a `DocSet`. Whenever a document is changed locally,
+// call `setDoc()` on the docSet. The connection registers a callback on the docSet, and it figures
+// out whenever there are changes that need to be sent to the remote peer.
+//
+// knownClock is the most recent VClock that we think the peer has (either because they've told us
+// that it's their clock, or because it corresponds to a state we have sent to them on this
+// connection). Thus, everything more recent than knownClock should be sent to the peer.
```

Comment:
Style - the names on knownClock and advertisedClock feel a little confusing to me since one is theirs and one is ours. Maybe they should be "theirClock" and "ourClock"? Or more explictly "lastHeardClock" and "lastSentClock"? I feel like the optimistic clock updates are likely to confuse someone very badly at some point because of network channel loss.

Reply:
Ok, renamed in 876e8c1.

What do you mean with optimistic clock updates? Updating `theirClock` when we send a message? I think that's important so that we don't double-send the same message unnecessarily. I am assuming that the channel is TCP-like, i.e. it won't drop random messages in the middle of a session, but just close the connection if messages are dropped and can't be retransmitted.